src/dmd/objc_glue.d

```diff
@@ -39,273 +37,343 @@ import dmd.backend.type;
+        Type tret = tf.next;
+        *ec = el_var(Symbols.getMsgSend(tret, ehidden !is null));
+    }
 
-    // clear tables
-    objc_smethVarNameTable = initStringTable(objc_smethVarNameTable);
-    objc_smethVarRefTable = initStringTable(objc_smethVarRefTable);
+    override void setupEp(elem* esel, elem** ep, int leftToRight)
+    {
+        if (esel)
+        {
+            // using objc-style "virtual" call
+            // add hidden argument (second to 'this') for selector used by dispatch function
+            if (leftToRight)
+                *ep = el_param(esel, *ep);
+            else
+                *ep = el_param(*ep, esel);
+        }
+    }
 
-    // also wipe out segment numbers
-    for (int s = 0; s < ObjcSegment.SEG_MAX; ++s)
-        objc_segList[s] = 0;
+    override void generateModuleInfo()
+    {
+        if (Symbols.hasSymbols)
+            Symbols.getModuleInfo();
+    }
 }
 
-Symbol *objc_getCString(const(char)* str, size_t len, const(char)* symbolName, ObjcSegment segment)
+struct Segments
 {
-    objc_hasSymbols = true;
-
-    // create data
-    scope dtb = new DtBuilder();
-    dtb.nbytes(cast(uint)(len + 1), str);
-
-    // find segment
-    int seg = objc_getSegment(segment);
-
-    // create symbol
-    Symbol *s;
-    s = symbol_name(symbolName, SCstatic, type_allocn(TYarray, tstypes[TYchar]));
-    s.Sdt = dtb.finish();
-    s.Sseg = seg;
-    return s;
-}
+    enum Id
+    {
+        cString,
+        imageInfo,
+        methodName,
+        moduleInfo,
+        selectorRefs
+    }
 
-Symbol *objc_getMethVarName(const(char)* s, size_t len)
-{
-    objc_hasSymbols = true;
+    private
+    {
+        __gshared static int[segmentData.length] segments;
+
+        __gshared static Segments[__traits(allMembers, Id).length] segmentData = [
+            Segments("__cstring", "__TEXT", S_CSTRING_LITERALS),
+            Segments("__objc_imageinfo", "__DATA", S_REGULAR | S_ATTR_NO_DEAD_STRIP),
+            Segments("__objc_methname", "__TEXT", S_CSTRING_LITERALS),
+            Segments("__objc_classlist", "__DATA", S_REGULAR | S_ATTR_NO_DEAD_STRIP),
+            Segments("__objc_selrefs", "__DATA", S_ATTR_NO_DEAD_STRIP | S_LITERAL_POINTERS)
+        ];
+
+        const(char)* sectionName;
+        const(char)* segmentName;
+        int flags;
+        int alignment = 3;
+    }
 
-    StringValue *sv = objc_smethVarNameTable.update(s, len);
-    Symbol *sy = cast(Symbol *) sv.ptrvalue;
-    if (!sy)
+    static int opIndex(Id id)
     {
-        __gshared size_t classnamecount = 0;
-        char[42] namestr;
-        sprintf(namestr.ptr, "L_OBJC_METH_VAR_NAME_%lu", classnamecount++);
-        sy = objc_getCString(s, len, namestr.ptr, ObjcSegment.SEGmethname);
-        sv.ptrvalue = sy;
+        auto segmentsPtr = segments.ptr;
+
+        if (segmentsPtr[id] != 0)
+            return segmentsPtr[id];
+
+        foreach (i, seg ; segmentData)
+        {
+            segmentsPtr[i] = MachObj.getsegment(
```

Comment:
Do you really need to use `MachObj` here? I thought that dmd encapsulated all object related things into `Obj`.

Reply:
No, I forgot to change that. That's also why the tests are failing.

Reply:
@ibuclaw  Ok, I see now that `getsegment` is declared in the subclasses, like `MachObj` and `ElfObj`, and not in the base class `Obj`. `getsegment` has different signatures for each object format so it looks like it cannot be moved to the base class. Any suggestions?

Reply:
As discussed, best course of immediate action would be to use `version(OSX)` to enclose the call to `MachObj` behind.  The `else` branch can just be a runtime `assert(0)` as no other target platform of dmd will get this far.

Reply:
That's a good idea.

Reply:
Done.